### PR TITLE
Module6: Routing

### PR DESCRIPTION
**Pull request by nkorczyk**
Changes proposed in this PR:

- Added `react-router`, `react-router-dom`.
- Routing added.
- By default user lands on a new page with `No Films found` state.
- When user clicks on a film item, redirect him to `localhost/film/id`.
- When user performs a new search, you should redirect him to: `localhost/search/Search%20Query`
- Added handling `404 - Page Not Found` in caso fo invalid URL.
- Added loading a movies with selected `id` & movies with the same `genre` are loaded.
- Updated unit tests.

Please check and review.

### CODE REVIEW

**Review by MichalWierzba — COMMENTED** [review]
Code review was done offline, during our meeting, and resulted in 21d34bcbd1ed3886da61893bb69d62b0f48561f9

Please finish refactoring we agreed on. Main points:
- add missing and update existing propTypes
- todo comments that currently exist in code
- consider changing naming convention of event handlers
  - use `on` prefix for prop names
  - use `handle` prefix for handler functions
  - example: `<SomeComponent onSearch={this.handleSearch} />`
- avoid passing props multiple levels down in components tree

File: src/App.jsx [hunk 1]
```diff
@@ -2,59 +2,44 @@ import React, { Component } from 'react';
...
 import MoviePage from './components/MoviePage';
+import NotFoundPage from "./components/NotFoundPage";
+import { BrowserRouter, Route, Switch } from "react-router-dom";
+import PropTypes from 'prop-types';
 
 class App extends Component {
 
   componentDidMount() {
-    this.props.loadMovies();
+    this.props.clearStore();
```
Comment (MichalWierzba) [inline src/App.jsx]: Is clearing store actually needed?

File: src/actions/actionCreator.js [hunk 2]
```diff
@@ -33,11 +34,48 @@ export const loadMoviesRequest = () => ({
...
+export const selectMovie = (movie) => ({
+  type: ACTION_TYPES.SELECT_MOVIE,
+  movie
+});
+
+export const clearStore = () => ({
+  type: ACTION_TYPES.CLEAR_STORE,
+});
+
+export const persistLastSearchPhrase = (lastSearchPhrase) => {
```
Comment (MichalWierzba) [inline src/actions/actionCreator.js:42]: Let's discuss if this persisting is needed.

File: src/actions/actionCreator.js [hunk 3]
```diff
@@ -33,11 +34,48 @@ export const loadMoviesRequest = () => ({
...
+    lastSearchPhrase
+  });
+};
+
+export const LoadMovieDetailsSuccess = (movie) => ({
+  type: ACTION_TYPES.LOAD_MOVIE_DETAILS_SUCCESS,
+  movie
+});
+
+export const LoadMovieSimilarGenre = (movies) => ({
```
Comment (MichalWierzba) [inline src/actions/actionCreator.js]: Please update letter-case. All your action creators has camelCase

File: src/actions/actionCreator.js [hunk 4]
```diff
@@ -33,11 +34,48 @@ export const loadMoviesRequest = () => ({
...
   const searchBy = `&searchBy=${state.search.searchby === "title" ? "title" : "genres"}`;
-  const phrase = `?search=${state.search.phrase}`;
+  const phrase = `?search=${state.search.lastSearchPhrase}`;
+  const order = "&sortOrder=desc";
+  const limit = "&limit=15";
+
+  return `${url}${phrase}${searchBy}${order}${limit}`;
+}
+
+export const buildUrlByGenre = (getState, genre) => {
```
Comment (MichalWierzba) [inline src/actions/actionCreator.js]: Common part of URL building should be extracted to separate function

File: src/actions/actionCreator.js [hunk 5]
```diff
@@ -33,11 +34,48 @@ export const loadMoviesRequest = () => ({
...
+  const state = getState();
+  const url = baseURL;
+  const searchBy = `&searchBy=${genre}`;
+  const phrase = `?search=${state.search.lastSearchPhrase}`;
```
Comment (MichalWierzba) [inline src/actions/actionCreator.js]: It should be genre value to search/phrase

**Review by MichalWierzba — COMMENTED** [review]
Review in process. Please check current comments

File: src/App.jsx (hunk 1 above)
Comment (nkorczyk) [reply src/App.jsx]: Removed. I don't use it any more.

File: src/actions/actionCreator.js (hunk 2 above)
Comment (nkorczyk) [reply src/actions/actionCreator.js:42]: Removed. No longer needed.

File: src/actions/actionCreator.js (hunk 3 above)
Comment (nkorczyk) [reply src/actions/actionCreator.js]: Fixed.

File: src/actions/actionCreator.js (hunk 4 above)
Comment (nkorczyk) [reply src/actions/actionCreator.js]: URL building method refactored.

File: src/actions/actionCreator.js (hunk 5 above)
Comment (nkorczyk) [reply src/actions/actionCreator.js]: Fixed.

**Comment by nkorczyk** [comment]
Requested changes have been applied.
Thank you for sharing good JavaScript code refactoring practices.

File: src/components/MovieDetails.jsx [hunk 6]
```diff
@@ -1,35 +1,57 @@
...
-        <h3 className="red-text text-darken-2 title-details">{movie.title}</h3>
-        <h5 className="white-text text-darken-2 avg-rating">{movie.vote_average}</h5>
-        <p className="white-text text-darken-2">{movie.genres.join(" & ")}</p>
-        <span className="white-text text-darken-2">{movie.release_date}</span>
-        <span className="white-text text-darken-2">{movie.runtime + "min"}</span>
-        <p className="white-text text-darken-2">{movie.overview}</p>
+class MovieDetails extends Component {
+
+  componentDidMount() {
+    let url = `${baseURL}/${this.props.match.params.id}`;
```
Comment (MichalWierzba) [inline src/components/MovieDetails.jsx]: I'd suggest moving this out from component to action itself - component does not have to know about this URL

File: src/App.jsx [hunk 7]
```diff
@@ -2,59 +2,39 @@ import React, { Component } from 'react';
...
-    }
-    return data;
-  };
-
-  return movies.sort((a, b) => {
-    return dataForSort(b[sortbyDateOrRating]) - dataForSort(a[sortbyDateOrRating]);
-  });
+App.propTypes = {
+  loadMovies: PropTypes.func,
+  selectMovie: PropTypes.func,
```
Comment (MichalWierzba) [inline src/App.jsx]: Are these props used in this component? Please check and update.

File: src/components/MovieDetails.jsx [hunk 8]
```diff
@@ -1,35 +1,57 @@
...
 
 MovieDetails.propTypes = {
   movie: PropTypes.object
 };
 
-export default MovieDetails;
+const mapDispatchToProps = { getMovie };
+
+const mapStateToProps = (state, ownProps) => ({
+  movie: state.movies.data.find(({ id }) => id === Number(ownProps.match.params.id)),
```
Comment (MichalWierzba) [inline src/components/MovieDetails.jsx]: I think `movie` prop is not used anymore and can be removed.

File: src/components/Item.jsx [hunk 9]
```diff
@@ -1,15 +1,20 @@
...
-  const genres = movie.genres.join(" & ");
+const Item = (props) => {
+
+  const { id, poster_path, overview, title, release_date } = props.movie;
+  const genres = props.movie.genres.join(" & ");
 
   return (
     <div>
-      <img src={movie.poster_path} className="thumbnails" alt={movie.overview} />
+      <Link to={'/film/' + id}>
```
Comment (MichalWierzba) [inline src/components/Item.jsx]: You can use template string here as well.

File: src/components/MoviePage.jsx [hunk 10]
```diff
@@ -1,23 +1,34 @@
...
-  return (
-    <div>
-        <MovieDetails movie={movies[0]} />
+class MoviePage extends Component {
+  render() {
+    const { movies } = this.props;
+    return (
+      <div>
+        <MovieDetails />
         <Results items={movies.length} />
```
Comment (MichalWierzba) [inline src/components/MoviePage.jsx]: `items` prop changed it's name and is already set inside `Results.jsx` file (see `mapStateToProps`)

File: src/components/MoviePage.jsx [hunk 11]
```diff
@@ -1,23 +1,34 @@
...
         <Content movies={movies} />
```
Comment (MichalWierzba) [inline src/components/MoviePage.jsx]: `movies` prop is already set inside `Content.jsx` file (see `mapStateToProps`)

File: src/reducers/movies.js [hunk 12]
```diff
@@ -19,6 +19,21 @@ function movies(state = initialState.movies, action) {
...
       };
+    case ACTION_TYPES.SELECT_MOVIE:
+      return {
+        ...state,
+        selectedMovie: action.movie
+      };
+    case ACTION_TYPES.LOAD_MOVIE_DETAILS_SUCCESS:
+      return {
+        ...state,
+        selectedMovie: [action.movie.data]
```
Comment (MichalWierzba) [inline src/reducers/movies.js]: I think there is no need to wrap `action.movie.data` in array.

**Review by MichalWierzba — APPROVED** [review]
Please do the last cleanup (according to comments) and it's 👍

File: src/App.jsx (hunk 7 above)
Comment (nkorczyk) [reply src/App.jsx]: Updated.

File: src/components/MovieDetails.jsx (hunk 8 above)
Comment (nkorczyk) [reply src/components/MovieDetails.jsx]: Removed.

File: src/components/Item.jsx (hunk 9 above)
Comment (nkorczyk) [reply src/components/Item.jsx]: Template string used.

File: src/components/MoviePage.jsx (hunk 10 above)
Comment (nkorczyk) [reply src/components/MoviePage.jsx]: Fixed.

File: src/components/MoviePage.jsx (hunk 11 above)
Comment (nkorczyk) [reply src/components/MoviePage.jsx]: Fixed.

File: src/reducers/movies.js (hunk 12 above)
Comment (nkorczyk) [reply src/reducers/movies.js]: Fixed. Array is not needed. Object is sufficient.

File: src/components/MovieDetails.jsx (hunk 6 above)
Comment (nkorczyk) [reply src/components/MovieDetails.jsx]: Fixed.

**Review by MichalWierzba — APPROVED** [review]

**Pull request merged**